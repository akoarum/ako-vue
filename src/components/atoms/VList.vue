<script lang="ts">
import Vue, { PropType, CreateElement, VNode } from 'vue'
import { ListItem } from '~/models/ListItem'

export default Vue.extend({
  props: {
    items: { type: Array as PropType<ListItem[]>, required: true },
    tag: { type: String, default: 'ul' },
    number: { type: Boolean }
  },
  computed: {
    styledClasses(): string[] {
      const classes: string[] = ['vList']

      if (this.number) classes.push('-number')

      return classes
    }
  },
  render(createElement: CreateElement): VNode {
    const Items = this.items.map(item => {
      const content = [item.label]

      return createElement(
        'li',
        { class: 'vList__item', key: item.id },
        content
      )
    })
    return createElement(
      this.tag,
      { class: this.styledClasses.join(' ') },
      Items
    )
  }
})
</script>

<style lang="scss" scoped>
.vList {
  @include media-min {
    font-size: $fontSize-pc;
  }

  @include media-max {
    font-size: $fontSize-sp;
  }
}

.vList.-number {
  counter-reset: number;
}

.vList__item {
  position: relative;
  padding-left: 1em;
  white-space: pre-line;
}

.-number .vList__item {
  counter-increment: number;
  padding-left: 1.2em;
}

.vList__item::before {
  position: absolute;
  left: 4px;
  content: '';
  border-radius: 50%;
  background: $color-theme;

  @include media-min {
    top: 0.55em;
    width: 5px;
    height: 5px;
  }

  @include media-max {
    top: 0.6em;
    width: 4px;
    height: 4px;
  }
}

.-number .vList__item::before {
  top: 0.05em;
  left: 0;
  width: auto;
  height: auto;
  border-radius: 0;
  background: transparent;
  content: counter(number) '.';
}

.vList__item + .vList__item {
  margin-top: 6px;
}
</style>
