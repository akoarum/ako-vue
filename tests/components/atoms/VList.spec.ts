import { shallowMount } from '@vue/test-utils'
import VList from '~/components/atoms/VList.vue'

describe('VList', () => {
  let wrapper: any

  beforeEach(() => {
    wrapper = shallowMount(VList, {
      propsData: {
        items: [
          { id: 1, label: 'リスト項目1' },
          { id: 2, label: 'リスト項目2' },
          { id: 3, label: 'リスト項目3' }
        ]
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('render', () => {
    it('何も指定しなければulになる', () => {
      expect(wrapper.is('ul')).toBe(true)
    })

    it('tagにolを指定すればolでレンダリングする', () => {
      wrapper.setProps({ tag: 'ol' })
      expect(wrapper.is('ol')).toBe(true)
    })
  })

  describe('computed', () => {
    it('styledClasses: propsに基づいてクラス配列を生成して返す', () => {
      expect(wrapper.vm.styledClasses).toEqual(['vList'])

      wrapper.setProps({ number: true })
      expect(wrapper.vm.styledClasses).toEqual(['vList', '-number'])
    })
  })
})
