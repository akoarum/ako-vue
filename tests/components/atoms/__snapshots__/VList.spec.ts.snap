// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VList initialized snapshot 1`] = `
<ul
  class="vList"
>
  <li
    class="vList__item"
  >
    リスト項目1
  </li>
  <li
    class="vList__item"
  >
    リスト項目2
  </li>
  <li
    class="vList__item"
  >
    リスト項目3
  </li>
</ul>
`;
